src/Core/Worker.php

```diff
@@ -0,0 +1,993 @@
+	}
+
+	/**
+	 * @brief Removes a workerqueue entry from the current process
+	 */
+	public static function unclaimProcess() {
+		$mypid = getmypid();
+
+		dba::update('workerqueue', array('executed' => NULL_DATE, 'pid' => 0), array('pid' => $mypid, 'done' => false));
+	}
+
+	/**
+	 * @brief Call the front end worker
+	 */
+	public static function callWorker() {
+		if (!Config::get("system", "frontend_worker")) {
+			return;
+		}
+
+		$url = System::baseUrl()."/worker";
+		fetch_url($url, false, $redirects, 1);
+	}
+
+	/**
+	 * @brief Call the front end worker if there aren't any active
+	 */
+	public static function executeIfIdle() {
+		if (!Config::get("system", "frontend_worker")) {
+			return;
+		}
+
+		// Do we have "proc_open"? Then we can fork the poller
+		if (function_exists("proc_open")) {
+			// When was the last time that we called the worker?
+			// Less than one minute? Then we quit
+			if ((time() - Config::get("system", "worker_started")) < 60) {
+				return;
+			}
+
+			set_config("system", "worker_started", time());
+
+			// Do we have enough running workers? Then we quit here.
+			if (self::tooMuchWorkers()) {
+				// Cleaning dead processes
+				self::killStaleWorkers();
+				get_app()->remove_inactive_processes();
+
+				return;
+			}
+
+			self::runCron();
+
+			logger('Call poller', LOGGER_DEBUG);
+			self::spawnWorker();
+			return;
+		}
+
+		// We cannot execute background processes.
+		// We now run the processes from the frontend.
+		// This won't work with long running processes.
+		self::runCron();
+
+		self::clearProcesses();
+
+		$workers = dba::fetch_first("SELECT COUNT(*) AS `processes` FROM `process` WHERE `command` = 'worker.php'");
+
+		if ($workers["processes"] == 0) {
+			self::callWorker();
+		}
+	}
+
+	/**
+	 * @brief Removes long running worker processes
+	 */
+	public static function clearProcesses() {
+		$timeout = Config::get("system", "frontend_worker_timeout", 10);
+
+		/// @todo We should clean up the corresponding workerqueue entries as well
+		$condition = array("`created` < ? AND `command` = 'worker.php'",
+				datetime_convert('UTC','UTC',"now - ".$timeout." minutes"));
+		dba::delete('process', $condition);
+	}
+
+	/**
+	 * @brief Runs the cron processes
+	 */
+	private static function runCron() {
+		logger('Add cron entries', LOGGER_DEBUG);
+
+		// Check for spooled items
+		self::add(PRIORITY_HIGH, "spool_post");
+
+		// Run the cron job that calls all other jobs
+		self::add(PRIORITY_MEDIUM, "cron");
+
+		// Run the cronhooks job separately from cron for being able to use a different timing
+		self::add(PRIORITY_MEDIUM, "cronhooks");
+
+		// Cleaning dead processes
+		self::killStaleWorkers();
+	}
+
+	public static function spawnWorker() {
+		$args = array("include/poller.php", "no_cron");
+		get_app()->proc_run($args);
+	}
+
+	/**
+	 * @brief Adds tasks to the worker queue
+	 *
+	 * @param (integer|array) priority or parameter array, $cmd atrings are deprecated and are ignored
```

Comment:
> $cmd atrings

?

Reply:
Yeah. You don't know the great "atrings"? ;-)